skimage/morphology/_max_tree.pyx

```diff
@@ -0,0 +1,459 @@
+        else:
+            # in this case the stop criterion is not fulfilled
+            # and this level is reconstructed entirely (flooding continues)
+            output[p] = image[p]
+
+    return
+
+# _build_max_tree is the main function. It allows to construct a max
+# tree representation of the image.
+cpdef void _build_max_tree(dtype_t[::1] image,
+                           DTYPE_BOOL_t[::1] mask,
+                           DTYPE_INT32_t[::1] structure,
+                           DTYPE_INT32_t[::1] shape,
+                           DTYPE_INT64_t[::1] parent,
+                           DTYPE_INT64_t[::1] sorted_indices
+                           ):
+    """Builds a max-tree.
+
+    Parameters
+    ----------
+
+    image : array of arbitrary type
+            The flattened image pixels.
+    mask : array of int
+        An array of the same shape as `image` where each pixel contains a
+        nonzero value if it is to be considered for the filtering.
+        NOTE: it is *essential* that the border pixels (those
+        with neighbors falling outside the volume) are all set to zero, or
+        segfaults could occur.
+    structure : array of int
+        A list of coordinate offsets to compute the raveled coordinates of each
+        neighbor from the raveled coordinates of the current pixel.
+    parent : array of int
+        First output: image of the same shape as the input image. The value
+        at each pixel is the parent index of this pixel in the max-tree
+        reprentation.
+    tree_order : array of int
+        Second output: list of length = number of pixels. Each element
+        corresponds to one pixel index in the image. It encodes the order
+        of elements in the tree: a parent of a pixel always comes before
+        the element itself. More formally: i < j implies that j cannot be
+        the parent of i.
+    """
+    cdef DTYPE_UINT64_t number_of_pixels = len(image)
+    cdef DTYPE_UINT64_t number_of_dimensions = len(shape)
+
+    cdef DTYPE_INT64_t i = 0
+    cdef DTYPE_INT64_t p = 0
+    cdef DTYPE_INT64_t root = 0
+    cdef DTYPE_INT64_t index = 0
+
+    cdef Py_ssize_t nneighbors = structure.shape[0]
+
+    cdef DTYPE_INT64_t[::1] zpar = parent.copy()
+
+    cdef np.ndarray[DTYPE_INT32_t, ndim = 2] points = offsets_to_points(structure, shape)
+
+    # initialization of the image parent.
+    for i in range(number_of_pixels):
+        parent[i] = -1
+        zpar[i] = -1
+
+    # traverse the array in reversed order (from highest value to lowest value)
+    for p in sorted_indices[::-1]:
+        parent[p] = p
+        zpar[p] = p
+
+        for i in range(nneighbors):
+
+            # get the ravelled index of the neighbor
+            index = p + structure[i]
+
+            if not mask[p]:
+                # in this case, p is at the border of the image.
+                # some neighbor point is not valid.
+                if not _is_valid_coordinate(p, points[i], shape):
```

Comment:
Ooooh, I see now. So the way that we solved this in e.g. watershed is by creating a bigger image and masking the border only. This way you can just ignore *all* border pixels. I don't know very well how that would affect performance here.

Reply:
The problem is that the function gives back an image that contains the parent indices of the ravelled image. With the strategy you employed in the WS function, all these indices would be shifted by offsets. For this function, I find this relatively messy. Therefore, I went for this strategy here (which is also not very elegant): if a pixel is not on the border, we can go ahead. If it is on the border, each neighbour must be checked whether it is in the image or not. But I am happy to change this if you have a good idea !

Reply:
Oooooohhh very good point! 👍